pkg/controllers/ratelimitpolicy/ratelimitpolicy_controller.go

```diff
@@ -0,0 +1,268 @@
+	MCTC_ATTRIBUTE_ANNOTATION_PREFIX = "kuadrant.io/attribute-"
+)
+
+// RateLimitPolicyReconciler reconciles a RateLimitPolicy object
+type RateLimitPolicyReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=kuadrant.io,resources=ratelimitpolicies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kuadrant.io,resources=ratelimitpolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=ratelimitpolicies/finalizers,verbs=update
+
+func (r *RateLimitPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	previous := &kuadrantapi.RateLimitPolicy{}
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, previous)
+	if err != nil {
+		if err := client.IgnoreNotFound(err); err == nil {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	rateLimitPolicy := previous.DeepCopy()
+
+	log.Log.V(3).Info("RateLimitPolicyReconciler Reconcile", "rateLimitPolicy", rateLimitPolicy)
+
+	if rateLimitPolicy.DeletionTimestamp != nil && !rateLimitPolicy.DeletionTimestamp.IsZero() {
+		controllerutil.RemoveFinalizer(rateLimitPolicy, RateLimitPolicyFinalizer)
+
+		err = r.Update(ctx, rateLimitPolicy)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	targetGateway, err := r.getTargetGateway(ctx, rateLimitPolicy)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = controllerutil.SetControllerReference(targetGateway, rateLimitPolicy, r.Scheme)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(rateLimitPolicy, RateLimitPolicyFinalizer) {
+		log.Log.Info("RateLimitPolicyFinalizer Reconcile: Added finalizer and set gateway owner", "rateLimitPolicy", rateLimitPolicy.Name)
+
+		controllerutil.AddFinalizer(rateLimitPolicy, RateLimitPolicyFinalizer)
+
+		err = r.Update(ctx, rateLimitPolicy)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	clusters, err := r.getClustersFromAnnotations(ctx, targetGateway)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	setSyncAnnotationsForClusters(rateLimitPolicy, clusters)
+	setPatchAnnotationsForClusters(rateLimitPolicy, clusters)
+
+	if !reflect.DeepEqual(rateLimitPolicy, previous) {
+		log.Log.V(3).Info("Updating RateLimitPolicy Spec", "rateLimitPolicySpec", rateLimitPolicy.Spec, "previousSpec", previous.Spec)
+		err = r.Update(ctx, rateLimitPolicy)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *RateLimitPolicyReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// Uncomment the following line adding a pointer to an instance of the controlled resource as an argument
+		For(&kuadrantapi.RateLimitPolicy{}).
```

Comment:
should we add a watch for the cluster secrets?

Reply:
Secrets are already being watched by the Gateway Reconciler, i just updated that to include RateLimitPolicies as a thing to update as required https://github.com/Kuadrant/multicluster-gateway-controller/pull/163/commits/bafcc3ab7399975217a561e885338f05e05ab58d